### PR TITLE
[model] Add Ideogram 4 support through Diffusers

### PR DESCRIPTION
**Pull request by nholmber**
## Summary

Adds xDiT support for Ideogram 4 using the native Diffusers 0.39 `Ideogram4Pipeline`, `Ideogram4Transformer2DModel`, and
`Ideogram4AttnProcessor` implementations.

The integration supports the official `ideogram-ai/ideogram-4-fp8` checkpoint, BF16 Diffusers checkpoints, USP/Ulysses, CFG parallelism, FSDP, optional parallel VAE decode, AITER BF16/FP8 attention, and torchao FP8/FP4 GEMMs.

**Requires:** Diffusers 0.39.0 or newer.

## Related Work

This supersedes [jackyhkust/xDiT#4](https://github.com/jackyhkust/xDiT/pull/4). That PR established the original Ideogram 4 xDiT path and several important multi-GPU, FP8-attention, prompt-broadcast, and compile-correctness fixes.

This version replaces the earlier bespoke model implementation with the native Diffusers Ideogram 4 API, uses xDiT's shared USP and attention-backend infrastructure, validates the existing AITER BF16/FP8 interfaces with `torch.compile`, and includes a new 48-case benchmark matrix.

## Features

- **Native Diffusers integration:** subclasses Diffusers' Ideogram 4 model, processor, and pipeline classes rather than vendoring the architecture
- **Parallelism:** USP/Ulysses, CFG parallel, Ulysses + CFG, FSDP, optional parallel VAE decoder
- **Quantization:** AITER FP8 attention, torchao FP8 GEMMs, AITER/torchao FP4 GEMMs
- **Official FP8 checkpoint loading:** converts fused QKV checkpoint keys to Diffusers' split Q/K/V layout and dequantizes weights into standard BF16 `nn.Linear` modules before runtime GEMM quantization
- **Prompt handling:** structured JSON prompts are passed through unchanged; plain-text prompts use the Ideogram prompt-enhancer head when available
- **Distributed prompt enhancement:** generates one enhanced prompt and broadcasts it across CFG and sequence-parallel CPU process groups
- **torch.compile:** both conditional and unconditional transformers compile with zero graph breaks
- **AITER attention validation:** the existing BF16 and FP8 interfaces compile without an Ideogram-specific wrapper when used with a compatible AITER build
- **Unpacked Ulysses communication:** uses the standard separate Q/K/V USP all-to-all path; this is 1.3-7.1% faster than materializing a packed QKV communication buffer

## Usage

```bash
# Single GPU, native 2048x2048, BF16 attention
python -m xfuser.cli \
  --model ideogram-ai/ideogram-4-fp8 \
  --prompt "A polished summer solstice festival poster" \
  --height 2048 --width 2048 --num_inference_steps 48 \
  --attention_backend AITER --use_torch_compile

# Single GPU, FP8 attention + FP4 GEMMs
python -m xfuser.cli \
  --model ideogram-ai/ideogram-4-fp8 \
  --prompt "A polished summer solstice festival poster" \
  --height 2048 --width 2048 --num_inference_steps 48 \
  --attention_backend AITER_FP8 --use_fp4_gemms \
  --use_torch_compile

# Four GPUs, Ulysses degree 2 + CFG parallel
python -m xfuser.cli \
  --model ideogram-ai/ideogram-4-fp8 \
  --prompt "$(cat benchmark/ideogram4/prompt_typography.json)" \
  --height 2048 --width 2048 --num_inference_steps 48 \
  --attention_backend AITER_FP8 --use_fp4_gemms \
  --ulysses_degree 2 --use_cfg_parallel --nproc_per_node 4 \
  --use_torch_compile
```

Ideogram 4 has 18 attention heads, so supported Ulysses degrees are `1`, `2`,
`3`, `6`, `9`, and `18`.

## Performance

Tested with `ideogram-ai/ideogram-4-fp8` on MI355X:

- 48 inference steps
- Seed 42
- Structured typography JSON prompt
- `torch.compile` enabled
- Five warmup calls
- Four measured iterations, with the first discarded
- PIL output, including VAE decode
- AITER commit `4a1cc773f34cbfc74387259e51262556ee38edd0`

### 1024x1024

| Parallel mode | BF16 | FP8 attention | FP8 GEMMs | FP4 GEMMs | FP8 attention + FP8 GEMMs | FP8 attention + FP4 GEMMs |
|---|---:|---:|---:|---:|---:|---:|
| U1 | 9.83s | 8.21s | 6.86s | 5.83s | 5.36s | 4.39s |
| U2 | 8.23s | 6.86s | 7.16s | 6.47s | 5.83s | 5.17s |
| CFG2 | 5.18s | 4.29s | 3.62s | 3.05s | 2.83s | **2.27s** |
| U2CFG2 | 4.32s | 3.62s | 3.73s | 3.31s | 3.03s | 2.63s |

Best result: **2.27s**, a **4.33x speedup** over U1 BF16.

### 2048x2048

| Parallel mode | BF16 | FP8 attention | FP8 GEMMs | FP4 GEMMs | FP8 attention + FP8 GEMMs | FP8 attention + FP4 GEMMs |
|---|---:|---:|---:|---:|---:|---:|
| U1 | 65.05s | 33.84s | 55.96s | 52.04s | 24.59s | 20.63s |
| U2 | 39.69s | 26.86s | 35.18s | 33.09s | 22.37s | 20.37s |
| CFG2 | 34.36s | 17.53s | 29.43s | 27.51s | 12.59s | 10.45s |
| U2CFG2 | 20.39s | 13.83s | 17.88s | 16.81s | 11.38s | **10.37s** |

Best result: **10.37s**, a **6.27x speedup** over U1 BF16.

### Packed vs Unpacked Ulysses

The initial implementation stacked Q, K, and V into one packed input all-to-all. Three separate Q/K/V collectives are faster for Ideogram 4 because the messages are bandwidth-bound and packing requires additional local tensor movement.

| Resolution | Mode | Minimum gain | Maximum gain |
|---|---|---:|---:|
| 1024x1024 | U2 / U2CFG2 | 1.35% | 6.69% |
| 2048x2048 | U2 / U2CFG2 | 1.76% | 7.12% |

## Architecture Notes

Ideogram 4 uses separate conditional and unconditional transformers rather than
running CFG as a doubled batch through one transformer.

- The conditional transformer processes text plus image tokens.
- The unconditional transformer processes image tokens only.
- CFG parallelism assigns one transformer branch to each CFG rank and gathers  one velocity tensor per denoising step.
- Ulysses chunks the compact text/image sequence across sequence-parallel ranks.
- The attention processor delegates to the existing `USP` interface, including   the configured AITER attention backend.

The FP8 checkpoint stores fused QKV weights and per-output-channel scales. The loader converts those keys to Diffusers' split `to_q`, `to_k`, and `to_v` modules, dequantizes to BF16, and leaves runtime FP8/FP4 GEMM selection to the existing xDiT quantization path.

## torch.compile and AITER

On gfx950, FP8 attention requires an AITER build containing the head-dimension 256 kernel `fmha_fwd_hd256_fp8_gfx950`, merged through ROCm/AITER PR 3732. This was validated against the official `amdsiloai/pytorch-xdit:v26.6` image by pulling the PR 3732 fully with its contents as it did not cherry-pick cleanly onto the version that the image ships with.

## Known Limitations

- Diffusers 0.39.0 or newer is required.
- AITER FP8 attention on gfx950 requires the HD256 FP8 FMHA kernel.
- Prompts in one batch must produce equal text-token lengths.
- PipeFusion and tensor parallelism are not implemented for Ideogram 4.
- The optional `outlines` package is recommended for schema-constrained prompt   enhancement. Without it, Diffusers runs prompt enhancement unconstrained.

### CODE REVIEW

File: docs/runner/runner.md [hunk 1]
```diff
@@ -102,6 +102,11 @@ Individual model classes that inherit from `xFuserModel`:
 | Qwen-Image-Edit | `Qwen-Image-Edit`, `Qwen/Qwen-Image-Edit`, `Qwen-Image-Edit-2509`, `Qwen/Qwen-Image-Edit-2509`, `Qwen-Image-Edit-2511`, `Qwen/Qwen-Image-Edit-2511` |
 | Krea2-Raw | `krea/krea-2-raw`, `krea/Krea-2-Raw`, `Krea-2-Raw` |
 | Krea2-Turbo | `krea/krea-2-turbo`, `krea/Krea-2-Turbo`, `Krea-2-Turbo` |
+| Ideogram 4 | `Ideogram-4`, `ideogram-ai/ideogram-v4`, `ideogram-ai/ideogram-4-nf4`, `ideogram-ai/ideogram-4-fp8` |
+
+Ideogram 4 requires Diffusers 0.39.0 or newer. Its 18 attention heads support Ulysses degrees `1`, `2`, `3`, `6`,
```
Comment (avjves) [inline docs/runner/runner.md]: This comment about Ideogram is maybe a bit out of place? There's a list in the main readme for the versions (though it's outdated..) and for the AITER comment we don't _really_ have a good place.

IMO we could remove this block entirely - the code itself can handle the diffusers version check.

File: xfuser/config/diffusers.py [hunk 2]
```diff
@@ -11,6 +11,7 @@
     "hunyuanvideo": "0.35.2",
```
Comment (avjves) [inline xfuser/config/diffusers.py]: A new min diffusion version check just landed in main. This file can now be removed. Can you add
```min_diffusers_version = "0.39.0"``` to the Ideogram xFuserModel?

File: xfuser/model_executor/models/runner_models/ideogram4.py [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+            strict=False,
+            assign=True,
+        )
+        _check_load_result("text_encoder", missing_keys, unexpected_keys)
+        text_encoder.eval()
+        log(f"Loaded {model_id}/text_encoder through the Ideogram FP8 converter.")
+        return text_encoder
+
+    def _load_model(self) -> DiffusionPipeline:
+        if not has_valid_diffusers_version("ideogram4"):
```
Comment (avjves) [inline xfuser/model_executor/models/runner_models/ideogram4.py]: These diffusers checks can now be removed (and the imports)

**Review by avjves — COMMENTED** [review]
Thanks! Left a couple of small comments :)

File: xfuser/model_executor/models/runner_models/ideogram4.py (hunk 3 above)
Comment (nholmber) [reply xfuser/model_executor/models/runner_models/ideogram4.py]: done

File: xfuser/config/diffusers.py (hunk 2 above)
Comment (nholmber) [reply xfuser/config/diffusers.py]: Done

File: docs/runner/runner.md (hunk 1 above)
Comment (nholmber) [reply docs/runner/runner.md]: Cleaned up

**Review by avjves — APPROVED** [review]
